### PR TITLE
Sort CSP filters differently (0, -1, 1, -2, 2, -3, ...)

### PR DESCRIPTION
**Pull request by cle1109**
I think the order of the CSP filters might not be correct. When we select `n_components` components, we usually pick the first, the last, the second, the second last, ... components until we have the desired number (@kazemakase and I do that in our SCoT implementation).

### CODE REVIEW

**Comment by kazemakase** [comment]
Maybe it's worth to point out that we did not arbitrarily choose this ordering. The largest and the smallest eigenvalues are associated with the two most discriminative components, followed by the second largest and second smallest eigenvalues, and so on...

I think the original code in MNE tries to achieve this sequence too, but it does not work as supposed :)

File: mne/decoding/csp.py [hunk 1]
```diff
@@ -129,8 +129,13 @@ def _fit(self, cov_a, cov_b):
         w_b = np.dot(np.dot(p, cov_b), p.T)
         # and solve it
         vals, vecs = linalg.eigh(w_a, w_b)
-        # sort vectors by discriminative power using eigen values
-        ind = np.argsort(np.maximum(vals, 1. / vals))[::-1]
+        # sort vectors by discriminative power using eigenvalues
+        ind = np.argsort(vals)[::-1]
+        vecs = vecs[:, ind]
+        # re-order (first, last, second, second last, third, ...)
+        ind = np.empty_like(ind)
```
Comment (kazemakase) [inline mne/decoding/csp.py]: No need to reallocate `ind`, because it is supposed to be completely filled below.

File: mne/decoding/csp.py [hunk 2]
```diff
@@ -129,8 +129,13 @@ def _fit(self, cov_a, cov_b):
...
+        ind[::2] = np.arange(0, np.ceil(len(ind)/2))
+        ind[1::2] = np.arange(len(ind) - 1, int(len(ind)/2) - 1, -1)
```
Comment (kazemakase) [inline mne/decoding/csp.py]: Did you miss the PEP8 spaces around ` / `?

Comment (cle1109) [inline mne/decoding/csp.py]: Spaces around `/` are not required - PEP8 says that [operators with the lowest priorities should be surrounded by spaces](https://www.python.org/dev/peps/pep-0008/#other-recommendations), so I guess in this case the style is OK.

File: mne/decoding/csp.py [hunk 3]
```diff
@@ -129,8 +129,12 @@ def _fit(self, cov_a, cov_b):
...
         # and solve it
         vals, vecs = linalg.eigh(w_a, w_b)
-        # sort vectors by discriminative power using eigen values
-        ind = np.argsort(np.maximum(vals, 1. / vals))[::-1]
+        # sort vectors by discriminative power using eigenvalues
+        ind = np.argsort(vals)[::-1]
+        vecs = vecs[:, ind]
+        # re-order (first, last, second, second last, third, ...)
+        ind[::2] = np.arange(0, np.ceil(len(ind)/2))
+        ind[1::2] = np.arange(len(ind) - 1, int(len(ind)/2) - 1, -1)
```
Comment (agramfort) [inline mne/decoding/csp.py]: int(len(ind)/2) -> len(ind) // 2

File: mne/decoding/csp.py [hunk 4]
```diff
@@ -129,8 +129,12 @@ def _fit(self, cov_a, cov_b):
         w_b = np.dot(np.dot(p, cov_b), p.T)
         # and solve it
         vals, vecs = linalg.eigh(w_a, w_b)
-        # sort vectors by discriminative power using eigen values
-        ind = np.argsort(np.maximum(vals, 1. / vals))[::-1]
+        # sort vectors by discriminative power using eigenvalues
+        ind = np.argsort(vals)[::-1]
+        vecs = vecs[:, ind]
+        # re-order (first, last, second, second last, third, ...)
+        ind[::2] = np.arange(0, np.ceil(len(ind)/2))
```
Comment (agramfort) [inline mne/decoding/csp.py]: np.ceil(len(ind) / 2.) -> int(np.ceil(len(ind) / 2.))

File: mne/decoding/csp.py (hunk 3 above)
Comment (agramfort) [inline mne/decoding/csp.py]: also maybe you can avoid taking 3 times len(ind) and introduce a variable n_vals ?

**Comment by agramfort** [comment]
you broke a test. cf. travis failure

**Comment by cle1109** [comment]
The Travis setup fails because it doesn't find `conda`. I ran `test_csp.py` on my local machine and the problem is resolved, so everything should be fine.

**Comment by agramfort** [comment]
@jaeilepp @Eric89GXL any idea about the travis / conda issue?

**Comment by jaeilepp** [comment]
No idea...

**Comment by Eric89GXL** [comment]
Yes:

https://github.com/ContinuumIO/anaconda-issues/issues/513

I'll fix it.

**Comment by Eric89GXL** [comment]
The AppVeyor issue is due to this, which we'll just have to wait for them to fix:

https://github.com/ContinuumIO/anaconda-issues/issues/511

So we'll have to ignore AppVeyor for a while :(

**Comment by cle1109** [comment]
Thanks @Eric89GXL - how do we proceed with this PR? Is there a way to trigger Travis manually or do we ignore it here as well?

**Comment by Eric89GXL** [comment]
Wait for https://github.com/mne-tools/mne-python/pull/2588 to land, then rebase, and it should trigger CIs properly

**Comment by Eric89GXL** [comment]
Well, Travis properly. CircleCI doesn't work currently, and AppVeyor has an issue we can't fix (Continuum IO needs to)

**Comment by Eric89GXL** [comment]
@cle1109 feel free to rebase and the CIs should run fine

**Comment by agramfort** [comment]
squashed, merged and back ported to 0.10 branch.

I also updated what's new

thanks @cle1109

**Pull request closed without merging**